xray/utils/downloadindexer.go

```diff
@@ -49,33 +55,123 @@ func DownloadIndexerIfNeeded(xrayManager *xray.XrayServicesManager) (string, err
+	tempDirPath := filepath.Join(indexerDirPath, tempIndexerDirName)
+
+	// Delete the temporary directory if it exists
+	tempDirExists, err := fileutils.IsDirExists(tempDirPath, false)
+	if err != nil {
+		return "", err
+	}
+	if tempDirExists {
+		err = os.RemoveAll(tempDirPath)
+		if err != nil {
+			return "", errorutils.CheckError(errors.New(fmt.Sprintf("Temporary download directory already exists, and can't be removed: %s\nRemove this directory manually and try again: %s", err.Error(), tempDirPath)))
+		}
+	}
+
+	// Delete all old indexers, but the two newest
+	err = deleteOldIndexers(indexerDirPath)
 	if err != nil {
 		return "", err
 	}
-	// Add execution premissions to the indexer
-	err = os.Chmod(indexerPath, 0777)
-	return indexerPath, err
-}
 
-func downloadIndexer(xrayManager *xray.XrayServicesManager, downloadDirPath string) error {
+	// Download the indexer from Xray to the temporary directory
 	url := fmt.Sprintf("%sapi/v1/indexer-resources/download/%s/%s", xrayManager.Config().GetServiceDetails().GetUrl(), runtime.GOOS, runtime.GOARCH)
 	downloadFileDetails := &httpclient.DownloadFileDetails{
 		DownloadPath:  url,
-		LocalPath:     downloadDirPath,
+		LocalPath:     tempDirPath,
 		LocalFileName: indexerFileName,
 	}
 	httpClientDetails := xrayManager.Config().GetServiceDetails().CreateHttpClientDetails()
-
 	resp, err := xrayManager.Client().DownloadFile(downloadFileDetails, "", &httpClientDetails, false)
 	if err == nil && resp.StatusCode != http.StatusOK {
 		body, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return errorutils.CheckError(errors.New(fmt.Sprintf("%s received when attempting to download %s. An error occurred while trying to read the body of the response: %s", resp.Status, url, err.Error())))
+			return "", errorutils.CheckError(errors.New(fmt.Sprintf("%s received when attempting to download %s. An error occurred while trying to read the body of the response: %s", resp.Status, url, err.Error())))
 		}
 		resp.Body.Close()
-		err = errorutils.CheckError(errors.New(fmt.Sprintf("%s received when attempting to download %s\n%s", resp.Status, url, body)))
+		return "", errorutils.CheckError(errors.New(fmt.Sprintf("%s received when attempting to download %s\n%s", resp.Status, url, body)))
+	}
+
+	// Add execution permissions to the indexer
+	indexerPath := filepath.Join(tempDirPath, indexerFileName)
+	err = os.Chmod(indexerPath, 0777)
+
+	indexerVersion, err := getIndexerVersion(indexerPath)
+	if err != nil {
+		return "", err
+	}
+	newDirPath := filepath.Join(indexerDirPath, indexerVersion)
+
+	// In case of a hot upgrade of Xray in progress, the version of the downloaded indexer might be different of the Xray version we got above,
+	// so the indexer we just downloaded may already exist.
+	newDirExists, err := fileutils.IsDirExists(newDirPath, false)
+	if err != nil {
+		return "", err
+	}
+	if newDirExists {
+		err = os.RemoveAll(tempDirPath)
+	} else {
+		err = os.Rename(tempDirPath, newDirPath)
```

Comment:
We're not allowed to use ```os.Rename```, because it will fail when running inside a docker image. Instead, you should use ```fileutils.MoveFile```.